README.md
Generate paper backups for Linux. Currently **command-line Linux only**.

**qr-backup** takes file(s), and outputs a "paper backup": a printable black-and-white pdf full of QR codes. To back up your file, print the PDF. The pile of paper in your hand is now a backup of the file.

If your file is lost, corrupted, deleted, etc, you can restore from your paper backup. qr-backup reads the [QR barcodes](https://en.wikipedia.org/wiki/QR_code) using your computer's webcam (or scanner) to get your file back.

## Example Backup
![Example Backup](docs/example.png)

## Features
- Restore without qr-backup installed (!)
- Restore using webcam
- Restore using scanner
- Clear, printed instructions on what the file is and how to restore
- Automatic compression
- 3KB/page on default settings. Single-digit MB backups of text are practical
- (Optional) Password protection
- (Optional) Print multiple copies for safety
- (Optional) Print smaller, denser codes to boost storage, up to 130KB/page
- (Restore with qr-backup only) Automatic redundancy. Lose up to 30% of pages or QR codes safely
- See [Command Line Options](docs/MAN.txt) for more

## [Install Guide](docs/INSTALL.md)
## [Command Line Options](docs/MAN.txt)

## How to Use
#### Making a backup
1. Run qr-backup on your file. On the Linux command-line, run `qr-backup <YOUR_FILE>`
2. This generates a black-and-white PDF (`<YOUR_FILE>.qr.pdf`)
3. Print the PDF on your printer

There are many command-line options available for advanced users. For a full list, read the [USAGE](docs/USAGE.md) doc online, or run `qr-backup --help` on your computer.

#### Restoring a paper backup
The restore process **does NOT require qr-backup**. It does require a command-line Linux computer.

(Option 1): Use qr-backup, if you have it.
- Webcam option
    1. Run `qr-backup --restore`
- Scanner option
    1. Scan images using your scanner
    2. Run `qr-backup --restore IMAGES`

(Option 2): Use the linux command line, if you lose qr-backup. Commands are provided in the PDF printout. You will need to install `zbar`.

## [FAQ](docs/FAQ.md)

### Paper Backups
- [Should I back up to paper?](docs/FAQ.md#should-i-back-up-to-paper)
- [What are the advantages of paper backups?](docs/FAQ.md#what-are-the-advantages-of-paper-backups)
- [How much of my backup can I lose and still restore?](docs/FAQ.md#how-much-of-my-backup-can-i-lose-and-still-restore)
- [Should I encrypt (docs/FAQpassword-protect) my backups?](#should-i-encrypt-password-protect-my-backups)
- [How can I protect my paper backup?](docs/FAQ.md#how-can-i-protect-my-paper-backup)

### Features
- [Do you support Windows?](docs/FAQ.md#do-you-support-windows)
- [Do you support mac/OS X?](docs/FAQ.md#do-you-support-macos-x)
- [How much data does this back up per page / why don't you back up more data per page?](docs/FAQ.md#how-much-data-does-this-back-up-per-page)
- [How do I back up more data per page?](docs/FAQ.md#how-do-i-back-up-more-data-per-page)
- [Why did you write qr-backup?](docs/FAQ.md#why-did-you-write-qr-backup)
- [Why doesn't the restore process require qr-backup?](docs/FAQ.md#why-doesnt-the-restore-process-require-qr-backup)
- [How exactly does the backup/restore process work?](docs/FAQ.md#how-exactly-does-the-backuprestore-process-work)
- [What are the design goals of qr-backup? / Why won't you add the feature I want?](docs/FAQ.md#what-are-the-design-goals-of-qr-backup)
- [What license is qr-backup released under?](docs/FAQ.md#what-license-is-qr-backup-released-under)

### Competition
- [What other paper backup projects exist?](docs/FAQ.md#what-other-paper-backup-projects-exist)
- [How does qr-backup compare to OllyDbg's Paperback?](docs/FAQ.md#how-does-qr-backup-compare-to-ollydbgs-paperback)

### Usage
- [How do I back up multiple files?](docs/FAQ.md#how-do-i-back-up-multiple-files)
- [Why does qr-backup restore to stdout? / why doesn't qr-backup extract tar files?](docs/FAQ.md#why-does-qr-backup-restore-to-stdout-rather-than-the-original-filename)

### Troubleshooting
- [My self-test is failing on Ubuntu](docs/FAQ.md#my-self-test-is-failing-on-ubuntu)
- [How do I find the maximum dimensions of my printer?](docs/FAQ.md#how-do-i-find-the-maximum-dimensions-of-my-printer)
- [When I print a page, part of it is cut off](docs/FAQ.md#when-i-print-a-page-part-of-it-is-cut-off)
- [When I print the backup, the last page is rotated](docs/FAQ.md#when-i-print-the-backup-the-last-page-is-rotated)


CHANGELOG
v1.1
- Added the `--restore` option so qr-backup will do the restore for you.
- Add erasure coding, which is turned on by default. This allows losing up
  to 30% of QR codes, but only if you are restoring using qr-backup.
- Added encryption. New command-line options:
  - `--encrypt PASS` uses a passphrase you supply.
  - `--encrypt generate` generates a random passphrase for you (printed on a
    separate page by default).
  - `--encrypt-print-passphrase` and `--no-encrypt-print-passphrase` turn the
    separate page on and off.
  - `--encrypt PASS` is also an option for restore.
- Automatic digital restore check added, which runs when you make a paper
  backup. Two tests are done testing restore from the digital .pdf. First, the
  linux command-line check is performed. Second qr-backup in --restore mode is
  used. Currently the check sometimes fails because zbar is unreliable, so
  failure only prints a warning.
- Backing up directories and multiple files is supporting using .tar .
  Tars are output as a single file during restore and not extracted.
- Added `-o`, `--output` option to give the name of an output file.
- Add `--shuffle`, `--no-shuffle`. This randomly orders QR codes. It's turned
  on by default is you use erasure coding. Losing random codes is better than
  losing consecutive codes for restore.
- Add `--num-copies`, which adds extra copies of every QR code for redundancy.
- Added `--instructions` option to pick how to print instructions. Options are
  as a cover sheet, on each page (the default), both, or neither.
- Added a english explanation of the restore process to the instructions. It's
  long so it's only on the cover sheet.
- Added the size (in bytes) of the file to the human-readable output.
- Added `--note` to add some user-chosen text to each page or the cover sheet.
- Format: Base64 encoding is now done within a QR code, not across QR codes.
  This is to support future erasure coding.
- Format: Codes are now labeled as "N01/50" rather than "01". This is to
  support multiple code types down the line, and to allow a progress bar during
  restore.
- Added --version, -V option
- Added manpage
- Added Makefile
- Bugfix (#6): Reading from stdin did not work correctly. Now it should.
- Bugfix (#7): The instructions are always full-page-width now.
- Bugfix (#15): Generated PDF is always identical (except the included date).
  However --encrypt is inherently non-deterministic for good reasons.
- Bugfix (#17): Improved the CLI restore to work with multiple consecutive
  restores
- Bugfix (#18): Support digital restore check when qr-backup is used in a
  pipeline, by keeping input and output in memory.
- Code cleanup: QR code content chunking vs actual QR graphic generation is
  split into two functions

v1.0
- Generate a PDF from any file
- Built-in compression (on by default)
- Uses base64 encoding, nothing, or '\n' to '\b' replacement as needed to avoid
  newlines in restore output.
- Long files are output as multiple QR codes.

- Default page size is 500px x 600px
- Default settings are 8px scale, version 10 QR codes, using binary encoding
  and M error correction. Output is 3K/side on default settings.
- Font is DevaVuSans (included)

- Restore is Linux CLI only, using zbar and coreutils.
- Release tested on Linux CLI only.
- Added README, FAQ, and USAGE.


test.py
#!/bin/python3
import hashlib
import math
import random
import subprocess
import sys
import time

def zeros(length):
    return b'\0'*length
def junk(length):
    r = random.Random(1)
    return bytes(r.getrandbits(8) for _ in range(length))

RED = '\033[91m'
GREEN = '\033[92m'
ENDC = '\033[0m'
def print_red(x):
    print(RED + x + ENDC)
def print_green(x):
    print(GREEN + x + ENDC)

DEFAULT_ARGS = ["--skip-checks", "-", "--backup-date", "2022-09-22"]
DEFAULT_RESTORE_ARGS = ["-"]

TESTS = [
    ("100b zeros",              zeros(100), [],                             2**0, [], 2**2),
    ("default options",         zeros(100), [
            "--dpi", "300",
            "--compress",
            "--erasure-coding",
            "--filename", "file",
            "--instructions", "page",
            "--num-copies 1",
            "--output", "-",
            "--page", "500", "600",
            "--qr-version", "10",
            "--scale", "5",
            "--shuffle",
                                             ],                             2**0, [], 2**2),
    ("--no-compress",           zeros(100), ["--no-compress"],              2**0, [], 2**2),
    ("--dpi",                   zeros(100), ["--dpi", "50"],                2**0, [], 2**3),
    ("--encrypt",               zeros(100), ["--encrypt", "PASSWORD"],      2**0, ["--encrypt", "PASSWORD"], 2**3),
    ("--encrypt-print-passphrase", zeros(100), ["--encrypt", "PASSWORD", "--encrypt-print-passphrase"], 2**0, ["--encrypt", "PASSWORD"], 2**3),
    ("--no-erasure-coding",     zeros(100), ["--no-erasure-coding"],        2**0, [], 2**3),
    ("--error-correction L",    zeros(100), ["--error-correction", "L"],    2**0, [], 2**2),
    ("--error-correction Q",    zeros(100), ["--error-correction", "Q"],    2**0, [], 2**2),
    ("--error-correction H",    zeros(100), ["--error-correction", "H"],    2**0, [], 2**2),
    ("--filename",              zeros(100), ["--filename", "TEST_FILENAME"],2**0, [], 2**2),
    ("--instructions both",     zeros(100), ["--instructions", "both"],     2**0, [], 2**3),
    ("--instructions cover",    zeros(100), ["--instructions", "cover"],    2**0, [], 2**3),
    ("--instructions none",     zeros(100), ["--instructions", "none"],     2**0, [], 2**2),
    ("--note (1)",              zeros(100), ["--note", "A"],                2**0, [], 2**2),
    ("--note (2)",              zeros(100), ["--note", "A", "--note", "B"], 2**0, [], 2**2),
    ("--num-copies",            zeros(100), ["--num-copies", "3"],          2**0, [], 2**2),
    ("--page",                  zeros(100), ["--page", "100", "200"],       2**0, [], 2**0),
    ("--qr-version",            zeros(100), ["--qr-version", "15"],         2**0, [], 2**2),
    ("--scale",                 zeros(100), ["--scale", "10"],              2**0, [], 2**2),
    ("--no-shuffle",            zeros(100), ["--no-shuffle"],               2**0, [], 2**2),
    ("1K zeros",                zeros(1000),[],                             2**0, [], 2**2),
    ("10K zeros",               zeros(10000), [],                           2**0, [], 2**2),
    ("100K zeros",              zeros(100000), [],                          2**0, [], 2**2),
    ("100b random",             junk(100),  [],                             2**0, [], 2**2),
    ("1K random",               junk(1000), [],                             2**0, [], 2**2),
    ("10K random",              junk(10000),[],                             2**3, [], 2**4),
    ("50K random",              junk(50000),[],                             2**5, [], 2**7),
    ("1K zeros, self-check",    zeros(1000),["--no-skip-checks"],           2**4, [], 2**2),
    ("1K random, self-check",   junk(1000), ["--no-skip-checks"],           2**4, [], 2**2),
]

BLESSED_OUTPUT = {
    '100b zeros':      'c6308f52da054b6e9eb5ce02e5fced8889ac80fb45c0c2bd60c1522ddce8ba14',
    'default options': 'c6308f52da054b6e9eb5ce02e5fced8889ac80fb45c0c2bd60c1522ddce8ba14', # This had better be the same!

    # Options need to be checked by hand when updating
    '--no-compress': '1d8ec11472c4709fda63f3112a408db2da6c50a112dac8b0b4e0548cbea673f5',
    '--dpi': 'd363989549c855e96e0dda855756829c6b1f8f11b8ffe854b02cf6104466c072',
    # Encryption is non-deterministic
    '--encrypt': None,
    "--encrypt-print-passphrase": None,
    '--no-erasure-coding': '6746f3ca2b75d7397eda1e63fab1167b5a4cde44fabe681d7a48b1486ed39cb6',
    '--error-correction L': '8ab1475050ec7ebdd1d20ca46eb4a7083031922ebd1a573dba0ca80abb25f65a',
    '--error-correction Q': 'ec943a14a116ca596bc8f58f3fe006e461d68dce788ea9926a822d659df292dc',
    '--error-correction H': '1aad991555e4162d9610d8e989600da5e4031a8399a7082afb032e26e9891200',
    '--filename': '8137004c57a545e386cc0459a4ac802b949601f3a640c8fd859415652dc1b0d4',
    '--instructions both': '725cb6be5d80a6e76b0166fea3205daa61c3a2db5404f4f2802c5208f0a157d1',
    '--instructions cover': '4878036208da2bb5b02a0aadc9473756f3e8e3898e3540cef75ea0f0dd186c99',
    '--instructions none': 'ed80f71717140037506abe1d3aea05d97278d64d97b906abfa6d9ff85286fe33',
    '--note (1)': 'cf904a8e4d87e3135ca81ce79c0e93dbdc4cb2775224426f99c27be5f5a18d78',
    '--note (2)': 'ed9100f706bff9984d11fc6a845074c47b721856484c87344c7cabfc1e1a2c83',
    '--num-copies': '19b89f0c9a770052b8ccdc00597056375306ae8e0278373bfd112406ff188f2e',
    '--page': '0696073fe4d8c5c4d1bda3f52036952860dc033abfd5ab002dbcde037b27d0c4',
    '--qr-version': '36f0ce65559ab37779a022c2492dbf1d029912fb9e740e92db053ee60ab6a1ed',
    '--scale': '3bda1a470226c115a0f3ad85f47748b71d40a1183c65e79ef29afa876bb86bc3',
    '--no-shuffle': 'ddbf172b473df1856a5015423446b4626f3dfdcaa644f290a4f84b06a5acb2a9',

    '1K zeros': 'd45df72505b649c41f6fb4a6aae31de1b6d6ee57f6d98141554bda186457b9b6',
    '10K zeros': 'ad3c0e8eccaedc88786b51b168573c6bf952f923766e6875e1579050d070516f',
    '100K zeros': 'bcefefc7441bcc341f4e2cda9a9f7f74b59bfa22b7a99bef4e6561999e465f66',
    '100b random': 'ce704c5e984ac7b54a78e5da19d69a02e3ffa505d3ae0733759433108f599fca',
    '1K random': '0541b29b91950263b3f0ff5936b1c994b3b576658257d79d7c49fa750a417619',
    '10K random': '9728f552c43d02d7a609528af78b84279152fcbeee0b06272d69b504a57ed037',
    '50K random': '93fcea86604b53e2c91e668303ef6a58c11a0782398171bb41c5a16370034015',
    '1K zeros, self-check': 'd45df72505b649c41f6fb4a6aae31de1b6d6ee57f6d98141554bda186457b9b6',
    '1K random, self-check': '0541b29b91950263b3f0ff5936b1c994b3b576658257d79d7c49fa750a417619',
}


def do_test(test, new_blessed):
    name, input_bytes, options, time_limit, restore_options, restore_time_limit = test
    expected_sha = BLESSED_OUTPUT.get(name, b'')

    failures = 0

    options = DEFAULT_ARGS + options
    qr_command = " ".join(["python3", "qr-backup"] + options)
    start = time.time()
    result = subprocess.run(qr_command, shell=True, capture_output=True, input=input_bytes)
    elapsed = time.time() - start
    output_bytes = result.stdout
    sha = hashlib.sha256(output_bytes).hexdigest()
    elapsed, power = math.ceil(elapsed), math.ceil(math.log(elapsed, 2))

    if expected_sha is None: # Some tests are non-deterministic (ones using encryption)
        pass
    elif sha == expected_sha:
        print_green("+b {} {}s".format(name, elapsed))
    else:
        print_red("-b {} {}s".format(name, elapsed))
        print("  command:", qr_command)
        print("  result:", sha, "!=", expected_sha)
        failures += 1
        failed = True
    if elapsed > time_limit*2:
        print_red("slow-b", name, "{}s, <2^{}".format(elapsed, power))
        failures += 1
        failed = True
    elif elapsed <= time_limit / 3:
        print("fast-b", name, "{}s, <2^{}".format(elapsed, power))
        pass

    restore_options = DEFAULT_RESTORE_ARGS + restore_options
    restore_command = " ".join(["python3", "qr-backup", "--restore"] + restore_options)
    start = time.time()
    result2 = subprocess.run(restore_command, shell=True, capture_output=True, input=output_bytes)
    elapsed = time.time() - start
    restored_bytes = result2.stdout
    elapsed, power = math.ceil(elapsed), math.ceil(math.log(elapsed, 2))

    if input_bytes == restored_bytes:
        print_green("+r {} {}s".format(name, elapsed))
        if expected_sha is not None and sha != expected_sha:
            new_blessed[name] = sha   
    else:
        print_red("-r {} {}s".format(name, elapsed))
        print("  command:", restore_command)
        #print(input_bytes, restored_bytes)
        failures += 1
    if elapsed > restore_time_limit*2:
        print_red("slow-r", name, "{}s, <2^{}".format(elapsed, power))
        failures += 1
    elif elapsed <= restore_time_limit / 3:
        print("fast-r", name, "{}s, <2^{}".format(elapsed, power))
        pass

    return failures, sha


if __name__ == "__main__":
    failures = 0
    new_blessed = {}
    for test in TESTS:
        new_failures, sha = do_test(test, new_blessed)
        failures += new_failures
        if False and new_failures > 0:
            with open("failure.bin", "wb") as f:
                f.write(test[1])
            print("exit on first failure")
            break

    if len(new_blessed) > 0:
        print("NEW BLESSED_OUTPUT = {")
        for k, v in new_blessed.items():
            print("{}: {},".format(repr(k), repr(v)))
        print("}")

    print("{} failures".format(failures))
    sys.exit(1 if failures > 0 else 0)


english_wordlist.py
# BIP 39 wordlist for mnemonic keys
words = "abandon ability able about above absent absorb abstract absurd abuse access accident account accuse achieve acid acoustic acquire across act action actor actress actual adapt add addict address adjust admit adult advance advice aerobic affair afford afraid again age agent agree ahead aim air airport aisle alarm album alcohol alert alien all alley allow almost alone alpha already also alter always amateur amazing among amount amused analyst anchor ancient anger angle angry animal ankle announce annual another answer antenna antique anxiety any apart apology appear apple approve april arch arctic area arena argue arm armed armor army around arrange arrest arrive arrow art artefact artist artwork ask aspect assault asset assist assume asthma athlete atom attack attend attitude attract auction audit august aunt author auto autumn average avocado avoid awake aware away awesome awful awkward axis baby bachelor bacon badge bag balance balcony ball bamboo banana banner bar barely bargain barrel base basic basket battle beach bean beauty because become beef before begin behave behind believe below belt bench benefit best betray better between beyond bicycle bid bike bind biology bird birth bitter black blade blame blanket blast bleak bless blind blood blossom blouse blue blur blush board boat body boil bomb bone bonus book boost border boring borrow boss bottom bounce box boy bracket brain brand brass brave bread breeze brick bridge brief bright bring brisk broccoli broken bronze broom brother brown brush bubble buddy budget buffalo build bulb bulk bullet bundle bunker burden burger burst bus business busy butter buyer buzz cabbage cabin cable cactus cage cake call calm camera camp can canal cancel candy cannon canoe canvas canyon capable capital captain car carbon card cargo carpet carry cart case cash casino castle casual cat catalog catch category cattle caught cause caution cave ceiling celery cement census century cereal certain chair chalk champion change chaos chapter charge chase chat cheap check cheese chef cherry chest chicken chief child chimney choice choose chronic chuckle chunk churn cigar cinnamon circle citizen city civil claim clap clarify claw clay clean clerk clever click client cliff climb clinic clip clock clog close cloth cloud clown club clump cluster clutch coach coast coconut code coffee coil coin collect color column combine come comfort comic common company concert conduct confirm congress connect consider control convince cook cool copper copy coral core corn correct cost cotton couch country couple course cousin cover coyote crack cradle craft cram crane crash crater crawl crazy cream credit creek crew cricket crime crisp critic crop cross crouch crowd crucial cruel cruise crumble crunch crush cry crystal cube culture cup cupboard curious current curtain curve cushion custom cute cycle dad damage damp dance danger daring dash daughter dawn day deal debate debris decade december decide decline decorate decrease deer defense define defy degree delay deliver demand demise denial dentist deny depart depend deposit depth deputy derive describe desert design desk despair destroy detail detect develop device devote diagram dial diamond diary dice diesel diet differ digital dignity dilemma dinner dinosaur direct dirt disagree discover disease dish dismiss disorder display distance divert divide divorce dizzy doctor document dog doll dolphin domain donate donkey donor door dose double dove draft dragon drama drastic draw dream dress drift drill drink drip drive drop drum dry duck dumb dune during dust dutch duty dwarf dynamic eager eagle early earn earth easily east easy echo ecology economy edge edit educate effort egg eight either elbow elder electric elegant element elephant elevator elite else embark embody embrace emerge emotion employ empower empty enable enact end endless endorse enemy energy enforce engage engine enhance enjoy enlist enough enrich enroll ensure enter entire entry envelope episode equal equip era erase erode erosion error erupt escape essay essence estate eternal ethics evidence evil evoke evolve exact example excess exchange excite exclude excuse execute exercise exhaust exhibit exile exist exit exotic expand expect expire explain expose express extend extra eye eyebrow fabric face faculty fade faint faith fall false fame family famous fan fancy fantasy farm fashion fat fatal father fatigue fault favorite feature february federal fee feed feel female fence festival fetch fever few fiber fiction field figure file film filter final find fine finger finish fire firm first fiscal fish fit fitness fix flag flame flash flat flavor flee flight flip float flock floor flower fluid flush fly foam focus fog foil fold follow food foot force forest forget fork fortune forum forward fossil foster found fox fragile frame frequent fresh friend fringe frog front frost frown frozen fruit fuel fun funny furnace fury future gadget gain galaxy gallery game gap garage garbage garden garlic garment gas gasp gate gather gauge gaze general genius genre gentle genuine gesture ghost giant gift giggle ginger giraffe girl give glad glance glare glass glide glimpse globe gloom glory glove glow glue goat goddess gold good goose gorilla gospel gossip govern gown grab grace grain grant grape grass gravity great green grid grief grit grocery group grow grunt guard guess guide guilt guitar gun gym habit hair half hammer hamster hand happy harbor hard harsh harvest hat have hawk hazard head health heart heavy hedgehog height hello helmet help hen hero hidden high hill hint hip hire history hobby hockey hold hole holiday hollow home honey hood hope horn horror horse hospital host hotel hour hover hub huge human humble humor hundred hungry hunt hurdle hurry hurt husband hybrid ice icon idea identify idle ignore ill illegal illness image imitate immense immune impact impose improve impulse inch include income increase index indicate indoor industry infant inflict inform inhale inherit initial inject injury inmate inner innocent input inquiry insane insect inside inspire install intact interest into invest invite involve iron island isolate issue item ivory jacket jaguar jar jazz jealous jeans jelly jewel job join joke journey joy judge juice jump jungle junior junk just kangaroo keen keep ketchup key kick kid kidney kind kingdom kiss kit kitchen kite kitten kiwi knee knife knock know lab label labor ladder lady lake lamp language laptop large later latin laugh laundry lava law lawn lawsuit layer lazy leader leaf learn leave lecture left leg legal legend leisure lemon lend length lens leopard lesson letter level liar liberty library license life lift light like limb limit link lion liquid list little live lizard load loan lobster local lock logic lonely long loop lottery loud lounge love loyal lucky luggage lumber lunar lunch luxury lyrics machine mad magic magnet maid mail main major make mammal man manage mandate mango mansion manual maple marble march margin marine market marriage mask mass master match material math matrix matter maximum maze meadow mean measure meat mechanic medal media melody melt member memory mention menu mercy merge merit merry mesh message metal method middle midnight milk million mimic mind minimum minor minute miracle mirror misery miss mistake mix mixed mixture mobile model modify mom moment monitor monkey monster month moon moral more morning mosquito mother motion motor mountain mouse move movie much muffin mule multiply muscle museum mushroom music must mutual myself mystery myth naive name napkin narrow nasty nation nature near neck need negative neglect neither nephew nerve nest net network neutral never news next nice night noble noise nominee noodle normal north nose notable note nothing notice novel now nuclear number nurse nut oak obey object oblige obscure observe obtain obvious occur ocean october odor off offer office often oil okay old olive olympic omit once one onion online only open opera opinion oppose option orange orbit orchard order ordinary organ orient original orphan ostrich other outdoor outer output outside oval oven over own owner oxygen oyster ozone pact paddle page pair palace palm panda panel panic panther paper parade parent park parrot party pass patch path patient patrol pattern pause pave payment peace peanut pear peasant pelican pen penalty pencil people pepper perfect permit person pet phone photo phrase physical piano picnic picture piece pig pigeon pill pilot pink pioneer pipe pistol pitch pizza place planet plastic plate play please pledge pluck plug plunge poem poet point polar pole police pond pony pool popular portion position possible post potato pottery poverty powder power practice praise predict prefer prepare present pretty prevent price pride primary print priority prison private prize problem process produce profit program project promote proof property prosper protect proud provide public pudding pull pulp pulse pumpkin punch pupil puppy purchase purity purpose purse push put puzzle pyramid quality quantum quarter question quick quit quiz quote rabbit raccoon race rack radar radio rail rain raise rally ramp ranch random range rapid rare rate rather raven raw razor ready real reason rebel rebuild recall receive recipe record recycle reduce reflect reform refuse region regret regular reject relax release relief rely remain remember remind remove render renew rent reopen repair repeat replace report require rescue resemble resist resource response result retire retreat return reunion reveal review reward rhythm rib ribbon rice rich ride ridge rifle right rigid ring riot ripple risk ritual rival river road roast robot robust rocket romance roof rookie room rose rotate rough round route royal rubber rude rug rule run runway rural sad saddle sadness safe sail salad salmon salon salt salute same sample sand satisfy satoshi sauce sausage save say scale scan scare scatter scene scheme school science scissors scorpion scout scrap screen script scrub sea search season seat second secret section security seed seek segment select sell seminar senior sense sentence series service session settle setup seven shadow shaft shallow share shed shell sheriff shield shift shine ship shiver shock shoe shoot shop short shoulder shove shrimp shrug shuffle shy sibling sick side siege sight sign silent silk silly silver similar simple since sing siren sister situate six size skate sketch ski skill skin skirt skull slab slam sleep slender slice slide slight slim slogan slot slow slush small smart smile smoke smooth snack snake snap sniff snow soap soccer social sock soda soft solar soldier solid solution solve someone song soon sorry sort soul sound soup source south space spare spatial spawn speak special speed spell spend sphere spice spider spike spin spirit split spoil sponsor spoon sport spot spray spread spring spy square squeeze squirrel stable stadium staff stage stairs stamp stand start state stay steak steel stem step stereo stick still sting stock stomach stone stool story stove strategy street strike strong struggle student stuff stumble style subject submit subway success such sudden suffer sugar suggest suit summer sun sunny sunset super supply supreme sure surface surge surprise surround survey suspect sustain swallow swamp swap swarm swear sweet swift swim swing switch sword symbol symptom syrup system table tackle tag tail talent talk tank tape target task taste tattoo taxi teach team tell ten tenant tennis tent term test text thank that theme then theory there they thing this thought three thrive throw thumb thunder ticket tide tiger tilt timber time tiny tip tired tissue title toast tobacco today toddler toe together toilet token tomato tomorrow tone tongue tonight tool tooth top topic topple torch tornado tortoise toss total tourist toward tower town toy track trade traffic tragic train transfer trap trash travel tray treat tree trend trial tribe trick trigger trim trip trophy trouble truck true truly trumpet trust truth try tube tuition tumble tuna tunnel turkey turn turtle twelve twenty twice twin twist two type typical ugly umbrella unable unaware uncle uncover under undo unfair unfold unhappy uniform unique unit universe unknown unlock until unusual unveil update upgrade uphold upon upper upset urban urge usage use used useful useless usual utility vacant vacuum vague valid valley valve van vanish vapor various vast vault vehicle velvet vendor venture venue verb verify version very vessel veteran viable vibrant vicious victory video view village vintage violin virtual virus visa visit visual vital vivid vocal voice void volcano volume vote voyage wage wagon wait walk wall walnut want warfare warm warrior wash wasp waste water wave way wealth weapon wear weasel weather web wedding weekend weird welcome west wet whale what wheat wheel when where whip whisper wide width wife wild will win window wine wing wink winner winter wire wisdom wise wish witness wolf woman wonder wood wool word work world worry worth wrap wreck wrestle wrist write wrong yard year yellow you young youth zebra zero zone zoo".split()
assert len(words) == 2048


test-page.py
#!/usr/bin/env python3
"""
Output a "test sheet" of scale and QR version sizes. Varying error correction (L/M/H/Q) might be worthwhile, but it's too many factors, so I kept it to the basics.

Add a symlink qr-backup -> qrbackup.py to use
"""
import PIL, PIL.Image, PIL.ImageDraw, PIL.ImageFont, PIL.ImageOps
import english_wordlist
import base64, datetime, gzip, hashlib, io, logging, math, os, random, subprocess, sys, textwrap, qrcode
assert sys.version_info >= (3,6), "Python 3.6 is required. Submit a patch removing f-strings to fix it, sucka!"


from qrbackup import *

# Defaults, not configurable
PAGE_W_POINTS, PAGE_H_POINTS = 500,600
DPI=300
ERROR_CORRECTION = M # default, 25%
QR_VERSIONS = [2,3,4,5,6,7,8,9,10,12,14,16,18,20,25,30,35,40]
SCALES = [1,2,3,4,5,6,7,8]

EXAMPLE_DATA = b"test"
PAGE_W_PIXEL, PAGE_H_PIXEL = math.floor(PAGE_W_POINTS/72.0*DPI), math.floor(PAGE_H_POINTS/72.0*DPI)

lookup = {}
samples = []
for qr_version in QR_VERSIONS:
    for scale in SCALES:
        # Figure out size of one QR code
        #digits, padding, example_qrs = qr_codes(EXAMPLE_DATA, error_correction=ERROR_CORRECTION, version=qr_version, scale=scale)
        #assert len(example_qrs) == 1, qr_version
        #qr_w_pixel, qr_h_pixel = example_qrs[0].make_image().copy().size
        qr_w_pixel = scale*qr_grid_size(qr_version)
        qr_h_pixel = qr_w_pixel
        padding=4*scale
        
        # Calculate number of QRs that can fit on a page
        qrs_per_page_w = max_units_with_padding(PAGE_W_PIXEL, qr_w_pixel, padding)
        qrs_per_page_h = max_units_with_padding(PAGE_H_PIXEL, qr_h_pixel, padding)
        qrs_per_page = qrs_per_page_w*qrs_per_page_h
        if qrs_per_page < 6:
            continue

        # Calculate characters per QR
        chars_per_qr = qr_size_chars(qr_version, QR_MODE, ERROR_CORRECTION) - 5

        # Calculate overall density
        density = chars_per_qr * qrs_per_page
        if density < 8000:
            continue
        print(density, ERROR_CORRECTION, qr_version, scale)

        # Make an image the encodes the result
        real_data = f"{density} {ERROR_CORRECTION} {qr_version} {scale}".encode("ascii")
        qr = qrcode.QRCode(
            version=qr_version,
            error_correction=ERROR_CORRECTION,
            box_size=scale,
        )
        qr.add_data(qrcode.util.QRData(real_data, mode=QR_MODE), optimize=0)
        qr.make(fit=False)

        # Store the result
        samples.append((qr_version, scale, qr_w_pixel+padding*2, density, qr))

samples.sort(key=lambda x: (x[2], x[3]))
rows = []

row = []
row_size = 0
for qr_version, scale, qr_w_pixel, density, qr in samples:
    if row_size + qr_w_pixel < PAGE_W_PIXEL:
        row.append(qr.make_image())
        row_size += qr_w_pixel
    else:
        print("row: ", len(row))
        rows.append(h_merge(row))
        row, row_size = [qr.make_image()], qr_w_pixel

pages = []

page = []
page_size = 0
for row in rows:
    if page_size + row.size[1] < PAGE_H_PIXEL:
        page.append(row)
        page_size += row.size[1]
    else:
        print("page: ", len(page))
        pages.append(v_merge(page))
        page, page_size = [row], row.size[1]

# Make everything uniform page size
for i, page in enumerate(pages):
    final_page = PIL.Image.new(mode="L", size=(PAGE_W_PIXEL, PAGE_H_PIXEL), color=255)
    final_page.paste(page, (0,0))
    pages[i] = final_page

print("num pages: ", len(pages))

pages[0].save("test.pdf", format="pdf", save_all=True, append_images=pages[1:], resolution=DPI, producer="qr-backup", title=f"qr-backup test print page")
